packages/medusa/src/services/new-totals.ts

```diff
@@ -0,0 +1,738 @@
+      : this.getShippingMethodTotals_.bind(this)
+
+    const shippingMethodsTotals: {
+      [lineItemId: string]: ShippingMethodTotals
+    } = {}
+    for (const shippingMethod of shippingMethods) {
+      shippingMethodsTotals[shippingMethod.id] = await calculationMethod(
+        shippingMethod,
+        {
+          includeTax,
+          calculationContext,
+          taxLines: shippingMethodsTaxLinesMap[shippingMethod.id],
+          discounts,
+          taxRate,
+        }
+      )
+    }
+
+    return shippingMethodsTotals
+  }
+
+  /**
+   * Calculate and return the shipping method totals
+   * @param shippingMethod
+   * @param includeTax
+   * @param calculationContext
+   * @param taxLines
+   * @param discounts
+   */
+  protected async getShippingMethodTotals_(
+    shippingMethod: ShippingMethod,
+    {
+      includeTax,
+      calculationContext,
+      taxLines,
+      discounts,
+    }: {
+      includeTax?: boolean
+      calculationContext: TaxCalculationContext
+      taxLines?: ShippingMethodTaxLine[]
+      discounts?: Discount[]
+    }
+  ) {
+    const totals: ShippingMethodTotals = {
+      price: shippingMethod.price,
+      original_total: shippingMethod.price,
+      total: shippingMethod.price,
+      subtotal: shippingMethod.price,
+      original_tax_total: 0,
+      tax_total: 0,
+      tax_lines: shippingMethod.tax_lines ?? [],
+    }
+
+    if (includeTax) {
+      totals.tax_lines = totals.tax_lines.length
+        ? totals.tax_lines
+        : (taxLines as ShippingMethodTaxLine[])
+
+      if (!totals.tax_lines) {
+        throw new MedusaError(
+          MedusaError.Types.UNEXPECTED_STATE,
+          "Tax Lines must be joined to calculate taxes"
+        )
+      }
+    }
+
+    const calculationContext_: TaxCalculationContext = {
+      ...calculationContext,
+      shipping_methods: [shippingMethod],
+    }
+
+    if (totals.tax_lines.length) {
+      const includesTax =
+        this.featureFlagRouter_.isFeatureEnabled(
+          TaxInclusivePricingFeatureFlag.key
+        ) && shippingMethod.includes_tax
+
+      totals.original_tax_total = await this.taxCalculationStrategy_.calculate(
+        [],
+        totals.tax_lines,
+        calculationContext_
+      )
+      totals.tax_total = totals.original_tax_total
+
+      if (includesTax) {
+        totals.subtotal -= totals.tax_total
+      } else {
+        totals.original_total += totals.original_tax_total
+        totals.total += totals.tax_total
+      }
+    }
+
+    const hasFreeShipping = discounts?.some(
+      (d) => d.rule.type === DiscountRuleType.FREE_SHIPPING
+    )
+
+    if (hasFreeShipping) {
+      totals.total = 0
+      totals.subtotal = 0
+      totals.tax_total = 0
+    }
+
+    return totals
+  }
+
+  /**
+   * Calculate and return the shipping method totals legacy using teh tax rate
+   * @param shippingMethod
+   * @param calculationContext
+   * @param taxLines
+   * @param discounts
```

Comment:
does not match method signature

Reply:
It is auto generated but it seams that tax rate is missing nice catch thanks 🙌